hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/checksum/TestContainerChecksumTreeManager.java

```diff
@@ -0,0 +1,164 @@
+  }
+
+  @Test
+  public void testWriteOnlyDeletedBlocksToFile() throws Exception {
+    List<Long> expectedBlocksToDelete = Arrays.asList(1L, 2L, 3L);
+    checksumManager.markBlocksAsDeleted(container, new TreeSet<>(expectedBlocksToDelete));
+
+    ContainerProtos.ContainerChecksumInfo checksumInfo = readFile();
+
+    assertEquals(CONTAINER_ID, checksumInfo.getContainerID());
+    assertEquals(expectedBlocksToDelete, checksumInfo.getDeletedBlocksList());
+    ContainerProtos.ContainerMerkleTree treeProto = checksumInfo.getDataMerkleTree();
+    assertEquals(0, treeProto.getDataChecksum());
+    assertTrue(treeProto.getBlockMerkleTreeList().isEmpty());
+  }
+
+  @Test
+  public void testDeletedBlocksPreservedOnTreeWrite() throws Exception {
+    List<Long> expectedBlocksToDelete = Arrays.asList(1L, 2L, 3L);
+    checksumManager.markBlocksAsDeleted(container, new TreeSet<>(expectedBlocksToDelete));
+    ContainerMerkleTree tree = buildTestTree();
+    checksumManager.writeContainerDataTree(container, tree);
+
+    ContainerProtos.ContainerChecksumInfo checksumInfo = readFile();
+
+    assertEquals(CONTAINER_ID, checksumInfo.getContainerID());
+    assertEquals(expectedBlocksToDelete, checksumInfo.getDeletedBlocksList());
+    assertTreesSortedAndMatch(tree.toProto(), checksumInfo.getDataMerkleTree());
+  }
+
+  @Test
+  public void testTreePreservedOnDeletedBlocksWrite() throws Exception {
+    ContainerMerkleTree tree = buildTestTree();
+    checksumManager.writeContainerDataTree(container, tree);
+    List<Long> expectedBlocksToDelete = Arrays.asList(1L, 2L, 3L);
+    checksumManager.markBlocksAsDeleted(container, new TreeSet<>(expectedBlocksToDelete));
+
+    ContainerProtos.ContainerChecksumInfo checksumInfo = readFile();
+
+    assertEquals(CONTAINER_ID, checksumInfo.getContainerID());
+    assertEquals(expectedBlocksToDelete, checksumInfo.getDeletedBlocksList());
+    assertTreesSortedAndMatch(tree.toProto(), checksumInfo.getDataMerkleTree());
+  }
+
+  private ContainerMerkleTree buildTestTree() throws Exception {
+    final long blockID1 = 1;
+    final long blockID2 = 2;
+    final long blockID3 = 3;
+    ChunkInfo b1c1 = buildChunk(0, ByteBuffer.wrap(new byte[]{1, 2, 3}));
+    ChunkInfo b1c2 = buildChunk(1, ByteBuffer.wrap(new byte[]{4, 5, 6}));
+    ChunkInfo b2c1 = buildChunk(0, ByteBuffer.wrap(new byte[]{7, 8, 9}));
+    ChunkInfo b2c2 = buildChunk(1, ByteBuffer.wrap(new byte[]{12, 11, 10}));
+    ChunkInfo b3c1 = buildChunk(0, ByteBuffer.wrap(new byte[]{13, 14, 15}));
+    ChunkInfo b3c2 = buildChunk(1, ByteBuffer.wrap(new byte[]{16, 17, 18}));
+
+    ContainerMerkleTree tree = new ContainerMerkleTree();
+    tree.addChunks(blockID1, Arrays.asList(b1c1, b1c2));
+    tree.addChunks(blockID2, Arrays.asList(b2c1, b2c2));
+    tree.addChunks(blockID3, Arrays.asList(b3c1, b3c2));
+
+    return tree;
+  }
+
+  private ContainerProtos.ContainerChecksumInfo readFile() throws IOException {
```

Comment:
Why not move this into the ChecksumTreeManager?

Reply:
The `ContainerChecksumInfo` file/proto is not meant to be interacted with directly. Updates and modifications to that file should go through the manager. This would not be clear if there was a public method in the manager to read the file directly.

Making the method public with `@VisibleForTesting` is also not ideal because the base directory used in each method is different so that would add an extra parameter.

Reply:
We are already exposing `writeContainerDataTree()`, Can we do something like that to test the `read() `functionality? I understand both `writeContainerDataTree()` and `markBlocksAsDeleted()` implicitly test `read()`. I have made this change as a part of #6864 to capture read latency. Let me know if it works in this case.

Reply:
Discussed offline:
- The read and write metrics are separate so the can still be tested even if the only public methods exposed are doing read-modify-write.
- GRPC API to move the tree between datanodes will operate at the file level and will not deserialize the protobuf.
- If a debug method is added to the client and we need to render the trees we can add such a method to hdds-common or similar package. That could also be invoked from this class if necessary.
  - This manager is designed to run in the datanode over a pool of containers and will not be suitable for client use anyways.